buildSrc/src/main/kotlin/protokt/v1/gradle/ProtoktBuild.kt

```diff
@@ -0,0 +1,240 @@
+
+private fun Project.linkGenerateProtoTasksAndIncludeGeneratedSource(sourceSetName: String, test: Boolean) {
+    val protoSourceSetRoot = if (test) "test" else "main"
+
+    val extension = project.extensions.getByType<ProtobufExtension>()
+
+    extension.generateProtoTasks.ofSourceSet(protoSourceSetRoot).forEach { genProtoTask ->
+        configureSourceSets(sourceSetName, protoSourceSetRoot, genProtoTask)
+        tasks.withType<AbstractKotlinCompile<*>> {
+            if ((test && name.contains("Test")) || (!test && !name.contains("Test"))) {
+                dependsOn(genProtoTask)
+            }
+        }
+    }
+}
+
+private fun Project.configureSourceSets(
+    sourceSetName: String,
+    protoSourceSetRoot: String,
+    genProtoTask: GenerateProtoTask
+) {
+    val kotlinExtension = extensions.getByName("kotlin") as ExtensionAware
+
+    @Suppress("UNCHECKED_CAST")
+    val sourceSets = kotlinExtension.extensions.getByName("sourceSets") as NamedDomainObjectContainer<KotlinSourceSet>
+
+    sourceSets.named(sourceSetName).configure {
+        kotlin.srcDir(genProtoTask.buildSourceDirectorySet())
+        the<SourceSetContainer>()
+            .getByName(protoSourceSetRoot)
+            .proto { resources.source(this) }
+    }
+}
+
+private fun GenerateProtoTask.buildSourceDirectorySet(): SourceDirectorySet {
+    val srcSetName = "generate-proto-$name"
+    val srcSet = objectFactory.sourceDirectorySet(srcSetName, srcSetName)
+    srcSet.srcDirs(
+        objectFactory
+            .fileCollection()
+            .builtBy(this)
+            .from(providerFactory.provider { outputSourceDirectoriesHack })
+    )
+    return srcSet
+}
+
+// The Kotlin JS plugin doesn't create main or test source sets
+private fun Project.configureJs(): Pair<String, String> {
+    // Can't call them "main" and "test" otherwise we'll get conflicting tasks 'processResources' and 'testClasses'
+    val mainSourceSet = the<SourceSetContainer>().create("jsMain")
+    val testSourceSet = the<SourceSetContainer>().create("jsTest")
+
+    val extension = project.extensions.getByType<ProtobufExtension>()
+
+    fun linkGenerateProtoTasksAndIncludeGeneratedSource(sourceSet: SourceSet, test: Boolean) {
+        val targetSourceSetName = if (test) "test" else "main"
+
+        extension.generateProtoTasks.ofSourceSet(sourceSet.name).forEach { genProtoTask ->
+            configureSourceSets(targetSourceSetName, sourceSet.name, genProtoTask)
+            the<SourceSetContainer>().getByName(sourceSet.name).proto { srcDir("src/$targetSourceSetName/proto") }
+            tasks.withType<AbstractKotlinCompile<*>> {
+                if ((test && name.contains("Test")) || (!test && !name.contains("Test"))) {
+                    dependsOn(genProtoTask)
+                }
+            }
+        }
+    }
+
+    linkGenerateProtoTasksAndIncludeGeneratedSource(mainSourceSet, false)
+    linkGenerateProtoTasksAndIncludeGeneratedSource(testSourceSet, true)
+
+    afterEvaluate {
```

Comment:
why afterEvaluate?

Reply:
I probably needed it long ago but it looks like I can get rid of it now.